python/hidet/runtime/compiled_task.py

```diff
@@ -147,6 +147,35 @@ def pick_best_candidate(self, inputs, outputs) -> int:
         return candidate_index
 
     def run_async(self, inputs):
+        from hidet import option, ir
+
+        if option.runtime_check():
+            symbol_map = {}
+            for i, (traced, new) in enumerate(zip(self.meta_data.inputs, inputs)):
+                if ir.data_type(traced.dtype) != new.dtype:
+                    raise RuntimeError(
+                        f"dtype mismatch at arg {i} between original: {traced.dtype} and new: {new.dtype}"
+                    )
+                traced_shape = traced.shape
+                concrete_shape = new.shape
+                if len(traced_shape) != len(concrete_shape):
+                    raise RuntimeError(
+                        f"Rank of input {i} not equal to original. ({len(concrete_shape)} vs. {len(traced_shape)})"
+                    )
+                for j, (orig_shape, new_shape) in enumerate(zip(traced_shape, concrete_shape)):
+                    if isinstance(orig_shape, int) and orig_shape != new_shape:
+                        raise RuntimeError(
+                            f'shape mismatch at dimension {j}, original: \
+                                           {orig_shape} vs. new: {new_shape}'
+                        )
+                    elif orig_shape not in symbol_map:
+                        symbol_map[orig_shape] = new_shape
+                    elif symbol_map[orig_shape] != new_shape:
+                        raise RuntimeError(
+                            f"There exists multiple instances of the same symbol {orig_shape}\
+                            with different values in inputs (ex: {symbol_map[orig_shape]} and {new_shape})"
+                        )
```

Comment:
Same logic as the CompiledGraph. Consider implement as a utility function and put it at the same module of `TensorSignature`.